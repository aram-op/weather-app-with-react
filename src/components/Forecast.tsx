import {ForecastListItem} from '../models/forecast-list-item.model';
import {fetchForecast} from '../services/weather.service';
import {useEffect, useState} from 'react';
import {FormattedForecast} from '../models/formatted-forecast.model';
import {Unit} from '../models/unit.model';
import '../styles/forecast.css';

function Forecast({unit, onError}: { unit: Unit, onError: (error: Error) => void }) {
    const [forecastList, setForecastList] = useState<FormattedForecast[]>([]);

    useEffect(() => {
        const getForecast = async () => {
            fetchForecast(unit)
                .then((forecast) => {
                    setForecastList(formatForecastData(forecast.list));
                })
                .catch(onError)
            ;
        }
        getForecast();
    }, []);

    const forecastItems = forecastList.map((item: any) => {
        return (
            <div className="forecast-item" key={item.day}>
                <p className="forecast-day">{item.day}</p>
                <img className="forecast-icon" src={`https://openweathermap.org/img/wn/${item.iconId}d@2x.png`}/>
                <div className="temp-container">
                    <span className="max-temp">{item.max}</span>
                    <span className="min-temp">{item.min}</span>
                </div>
            </div>
        );
    });

    return (
        <div className="forecast-container">
            {forecastItems}
        </div>
    );
}

function formatForecastData(list: ForecastListItem[]) {
    const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
    const forecastData = [];

    for (let i = 0; i < 4; i++) {
        const listIndex = 4 + 8 * i;
        const iconIds: number[] = [];
        let maxTemp = Math.round(list[listIndex].main.temp_max);
        let minTemp = Math.round(list[listIndex].main.temp_max);

        /**
         * The api returns forecast list for 40 hours.
         * The first 4 hours are from current day, thus the first 4 items in the list are skipped.
         * The last 4 hours are from the fifth day. They are skipped, because it represents only half of the day.
         */
        for (let j = listIndex + 1; j < listIndex + 8; j++) {
            const maxTempForHour = Math.round(list[j].main.temp_max);
            const minTempForHour = Math.round(list[j].main.temp_min);

            if (maxTempForHour > maxTemp) {
                maxTemp = maxTempForHour;
            }
            if (minTempForHour < minTemp) {
                minTemp = minTempForHour;
            }
            //IconId will be only the numeric part of the icon code.
            iconIds.push(parseInt(list[j].weather[0].icon.slice(0, 2)));
        }

        if (minTemp === -0) minTemp = 0;
        if (maxTemp === -0) maxTemp = 0;

        const day = days[new Date(list[listIndex].dt * 1000).getDay()];

        //Getting the worst case scenario icon id from the list.
        const iconId = Math.max(...iconIds) < 10 ? '0' + Math.max(...iconIds) : Math.max(...iconIds).toString();

        forecastData.push({min: minTemp, max: maxTemp, iconId: iconId, day: day});
    }
    return forecastData;
}

export default Forecast;