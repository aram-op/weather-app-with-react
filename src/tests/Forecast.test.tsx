import React from 'react';
import {render, screen, waitFor} from '@testing-library/react';
import Forecast from '../components/Forecast';
import {fetchForecast} from '../services/weather.service';
import {ForecastListItem} from '../models/forecast-list-item.model';

jest.mock('../services/weather.service');

const mockForecastData: ForecastListItem[] = [];
let mockTimestamp = 1733472000;
let mockMinTemp = 18;
let mockMaxTemp = 25;

for (let i = 0; i < 40; i++) {
    mockForecastData.push(
        {
            dt: mockTimestamp,
            main: {
                feels_like: 0,
                temp_min: mockMinTemp,
                temp_max: mockMaxTemp,
                pressure: 1015,
                sea_level: 1015,
                grnd_level: 933,
                humidity: 69,
                temp_kf: -1.11
            },
            weather: [
                {
                    id: 500,
                    main: 'Rain',
                    description: 'light rain',
                    icon: '10d'
                }
            ],
            clouds: {
                all: 100
            },
            wind: {
                speed: 0.62,
                deg: 349,
                gust: 1.18
            },
            visibility: 10000,
            pop: 0.32,
            rain: {
                h3: 0.26
            },
            sys: {
                pod: 'd'
            },
            dt_text: '2022-08-30 15:00:00'
        },
    );
    mockTimestamp += 10800;
    mockMinTemp++;
    mockMaxTemp++;
}


describe('Forecast component', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should displays the weather forecast', async () => {
        (fetchForecast as jest.Mock).mockResolvedValue({list: mockForecastData});

        render(<Forecast unit="metric" onError={jest.fn()}/>);

        await screen.findByText('Sat');

        expect(screen.getByText('Sat')).toBeInTheDocument();
        expect(screen.getByText('Sun')).toBeInTheDocument();
        expect(screen.getByText('Mon')).toBeInTheDocument();
        expect(screen.getByText('Tue')).toBeInTheDocument();
        expect(screen.getByText('23')).toBeInTheDocument();
        expect(screen.getByText('36')).toBeInTheDocument();
    });

    it('should call onError if error occurred while getting forecast data', async () => {
        const mockError = new Error('Failed to fetch forecast data');
        (fetchForecast as jest.Mock).mockRejectedValue(mockError);

        const onError = jest.fn();
        render(<Forecast unit="metric" onError={onError}/>);

        await waitFor(() => expect(onError).toHaveBeenCalledWith(mockError));
    });

    it('should match the snapshot', async () => {
        (fetchForecast as jest.Mock).mockResolvedValue({ list: mockForecastData });

        const { container } = render(<Forecast unit="metric" onError={jest.fn()} />);
        await screen.findByText('Sun');

        expect(container).toMatchSnapshot();
    });
});
