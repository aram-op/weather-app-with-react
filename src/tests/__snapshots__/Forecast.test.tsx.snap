// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast Component matches snapshot 1`] = `
<div>
  <div
    class="forecast-container"
  >
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Sat
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          36
        </span>
        <span
          class="min-temp"
        >
          23
        </span>
      </div>
    </div>
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Sun
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          44
        </span>
        <span
          class="min-temp"
        >
          31
        </span>
      </div>
    </div>
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Mon
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          52
        </span>
        <span
          class="min-temp"
        >
          39
        </span>
      </div>
    </div>
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Tue
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          60
        </span>
        <span
          class="min-temp"
        >
          47
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Forecast component should match snapshot 1`] = `
<div>
  <div
    class="forecast-container"
  >
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Sat
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          36
        </span>
        <span
          class="min-temp"
        >
          23
        </span>
      </div>
    </div>
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Sun
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          44
        </span>
        <span
          class="min-temp"
        >
          31
        </span>
      </div>
    </div>
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Mon
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          52
        </span>
        <span
          class="min-temp"
        >
          39
        </span>
      </div>
    </div>
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Tue
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          60
        </span>
        <span
          class="min-temp"
        >
          47
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Forecast component should match the snapshot 1`] = `
<div>
  <div
    class="forecast-container"
  >
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Sat
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          36
        </span>
        <span
          class="min-temp"
        >
          23
        </span>
      </div>
    </div>
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Sun
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          44
        </span>
        <span
          class="min-temp"
        >
          31
        </span>
      </div>
    </div>
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Mon
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          52
        </span>
        <span
          class="min-temp"
        >
          39
        </span>
      </div>
    </div>
    <div
      class="forecast-item"
    >
      <p
        class="forecast-day"
      >
        Tue
      </p>
      <img
        class="forecast-icon"
        src="https://openweathermap.org/img/wn/10d@2x.png"
      />
      <div
        class="temp-container"
      >
        <span
          class="max-temp"
        >
          60
        </span>
        <span
          class="min-temp"
        >
          47
        </span>
      </div>
    </div>
  </div>
</div>
`;
